addon/instance-initializers/prefetch.js

```diff
@@ -38,7 +41,15 @@ export function initialize(instance) {
       }
 
       // Run the prefetch hook if the route has one.
-      handlerInfo.handler._prefetched = handlerInfo.runSharedModelHook(transition, 'prefetch', [fullParams]);
+      if (!handlerInfo.handler && handlerInfo.handlerPromise) {
+        transition._handlerPromiseChain = transition._handlerPromiseChain.then(() => (
+          handlerInfo.handlerPromise.then((handler) => {
+            handler._prefetched = handlerInfo.runSharedModelHook(transition, 'prefetch', [fullParams]);
+          })
+        ));
+      } else {
+        handlerInfo.handler._prefetched = handlerInfo.runSharedModelHook(transition, 'prefetch', [fullParams]);
```

Comment:
Immediately available handlers still need to obey the hander promise chain. Take the test you wrote as an example. If the readiness of the handlers were reversed — `handler1` was deferred and `handler2` was immediately available — `handler2`'s prefetch hook would trigger this `else` block and run its prefetch hook before `handler1`'s.